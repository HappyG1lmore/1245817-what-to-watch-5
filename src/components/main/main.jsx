import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {filmsListPropTypes} from "../../common-prop-types";
import {connect} from "react-redux";
import GenresList from "../genres-list/genres-list";
import FilmsList from "../films-list/films-list";
import Header from "../header/header";
import Footer from "../footer/footer";
import MoreFilmsButton from "../more-films-button/more-films-button";
import {filteredFilmsSelector, genresFilterSelector} from "../../store/selectors";
import {MAX_AMOUNT_FILMS_PER_STEP} from "../../constants";
import {filmPropTypes} from "../../common-prop-types";
import FavoriteButton from "../favorite-button/favorite-button";

class Main extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      amountFilmsForRender: MAX_AMOUNT_FILMS_PER_STEP,
    };
    this.handleClickBtnShowMore = this.handleClickBtnShowMore.bind(this);
  }

  handleClickBtnShowMore() {
    let {amountFilmsForRender} = this.state;
    amountFilmsForRender += MAX_AMOUNT_FILMS_PER_STEP;
    this.setState({amountFilmsForRender});
  }

  componentDidUpdate(prevProps) {
    if (prevProps.filteredFilms !== this.props.filteredFilms) {
      this.setState({
        amountFilmsForRender: MAX_AMOUNT_FILMS_PER_STEP
      });
    }
  }

  render() {
    const {
      filteredFilms,
      filmsGenres,
      mainFilm,
      history
    } = this.props;

    const {amountFilmsForRender} = this.state;
    const filmsToRender = filteredFilms.slice(0, amountFilmsForRender);
    const isShowMoreVisible = amountFilmsForRender < filteredFilms.length;

    return (
      <>
        <section className="movie-card">
          <div className="movie-card__bg">
            <img
              src={mainFilm.background}
              alt={mainFilm.title}
            />
          </div>

          <h1 className="visually-hidden">WTW</h1>

          <Header
            className={`movie-card__head`}
          />

          <div className="movie-card__wrap">
            <div className="movie-card__info">
              <div className="movie-card__poster">
                <img
                  src={mainFilm.poster}
                  alt="The Grand Budapest Hotel poster"
                  width="218"
                  height="327"
                />
              </div>

              <div className="movie-card__desc">
                <h2 className="movie-card__title">{mainFilm.title}</h2>
                <p className="movie-card__meta">
                  <span className="movie-card__genre">{mainFilm.genre}</span>
                  <span className="movie-card__year">{mainFilm.year}</span>
                </p>

                <div className="movie-card__buttons">
                  <button className="btn btn--play movie-card__button" type="button"
                    onClick={() => {
                      history.push(`/player/${mainFilm.id}`);
                    }}>
                    <svg viewBox="0 0 19 19" width="19" height="19">
                      <use xlinkHref="#play-s"/>
                    </svg>
                    <span>Play</span>
                  </button>
                  <FavoriteButton
                    isFavorite={mainFilm.isFavorite}
                    id={mainFilm.id}
                  />
                </div>
              </div>
            </div>
          </div>
        </section>

        <div className="page-content">
          <section className="catalog">
            <h2 className="catalog__title visually-hidden">Catalog</h2>

            <ul className="catalog__genres-list">
              <GenresList filmGenres={filmsGenres}/>
            </ul>

            <div className="catalog__movies-list">
              <FilmsList filmsList={filmsToRender}/>
            </div>

            <MoreFilmsButton
              onClick={this.handleClickBtnShowMore}
              isVisible={isShowMoreVisible}
            />

          </section>

          <Footer />
        </div>
      </>
    );
  }
}

Main.propTypes = {
  mainFilm: filmPropTypes,
  filteredFilms: filmsListPropTypes,
  filmsGenres: PropTypes.arrayOf(PropTypes.string),
  history: PropTypes.object,
};

const mapStateToProps = (state) => {
  return {
    filteredFilms: filteredFilmsSelector(state.films),
    filmsGenres: genresFilterSelector(state.films),
    mainFilm: state.film.promoFilm
  };
};

export {Main};
export default connect(mapStateToProps)(Main);
