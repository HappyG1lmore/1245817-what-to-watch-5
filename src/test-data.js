const filmsRaw = [{
  "name": `Bronson`,
  "poster_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
  "preview_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
  "background_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
  "background_color": `#73B39A`,
  "description": `A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.`,
  "rating": 3.6,
  "scores_count": 109661,
  "director": `Nicolas Winding Refn`,
  "starring": [`Tom Hardy`, `Kelly Adams`, `Luing Andrews`],
  "run_time": 92,
  "genre": `Action`,
  "released": `2008`,
  "id": 1,
  "is_favorite": false,
  "video_link": `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
  "preview_video_link": `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  "name": `Pulp Fiction`,
  "poster_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg`,
  "preview_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg`,
  "background_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg`,
  "background_color": `#795433`,
  "description": `The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.`,
  "rating": 1.5,
  "scores_count": 1635992,
  "director": `Quentin Tarantino`,
  "starring": [`John Travolta`, `Uma Thurman`, `Samuel L. Jackson`],
  "run_time": 153,
  "genre": `Crime`,
  "released": `1994`,
  "id": 2,
  "is_favorite": false,
  "video_link": `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
  "preview_video_link": `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  "name": `Matrix`,
  "poster_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/matrix.jpg`,
  "preview_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/matrix.jpg`,
  "background_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/matrix.jpg`,
  "background_color": `#B9B27E`,
  "description": `A computer hacker learns from mysterious rebels about the true nature of his reality and his role in the war against its controllers.`,
  "rating": 4.4,
  "scores_count": 1503092,
  "director": `Wachowski Brothers`,
  "starring": [`Keanu Reeves`, `Laurence Fishburne`, `Carrie-Anne Moss`],
  "run_time": 136,
  "genre": `Action`,
  "released": `1999`,
  "id": 4,
  "is_favorite": false,
  "video_link": `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
  "preview_video_link": `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}];

const films = [{
  "title": `Bronson`,
  "poster": `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
  "frame": `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
  "background": `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
  "backgroundColor": `#73B39A`,
  "description": `A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.`,
  "ratingText": `Normal`,
  "rating": 3.6,
  "ratings": `109661`,
  "director": `Nicolas Winding Refn`,
  "starring": [`Tom Hardy`, `Kelly Adams`, `Luing Andrews`],
  "runtime": 92,
  "genre": `Action`,
  "year": `2008`,
  "id": 1,
  "isFavorite": false,
  "video": `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
  "previewVideoLink": `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  "title": `Pulp Fiction`,
  "poster": `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg`,
  "frame": `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg`,
  "background": `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg`,
  "backgroundColor": `#795433`,
  "description": `The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.`,
  "ratingText": `Bad`,
  "rating": 1.5,
  "ratings": `1635992`,
  "director": `Quentin Tarantino`,
  "starring": [`John Travolta`, `Uma Thurman`, `Samuel L. Jackson`],
  "runtime": 153,
  "genre": `Crime`,
  "year": `1994`,
  "id": 2,
  "isFavorite": false,
  "video": `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
  "previewVideoLink": `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  "title": `Matrix`,
  "poster": `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/matrix.jpg`,
  "frame": `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/matrix.jpg`,
  "background": `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/matrix.jpg`,
  "backgroundColor": `#B9B27E`,
  "description": `A computer hacker learns from mysterious rebels about the true nature of his reality and his role in the war against its controllers.`,
  "ratingText": `Normal`,
  "rating": 4.4,
  "ratings": `1503092`,
  "director": `Wachowski Brothers`,
  "starring": [`Keanu Reeves`, `Laurence Fishburne`, `Carrie-Anne Moss`],
  "runtime": 136,
  "genre": `Action`,
  "year": `1999`,
  "id": 4,
  "isFavorite": false,
  "video": `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
  "previewVideoLink": `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}];

const film = {
  "title": `Pulp Fiction`,
  "poster": `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg`,
  "frame": `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg`,
  "background": `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg`,
  "backgroundColor": `#795433`,
  "description": `The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.`,
  "rating": 1.5,
  "ratings": `1635992`,
  "director": `Quentin Tarantino`,
  "starring": [`John Travolta`, `Uma Thurman`, `Samuel L. Jackson`],
  "runtime": 153,
  "genre": `Crime`,
  "year": `1994`,
  "id": 2,
  "isFavorite": false,
  "video": `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
  "previewVideoLink": `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
};

const reviews = [{
  "id": 1,
  "user": {
    "id": 19,
    "name": `Christina`
  },
  "rating": 3.3,
  "comment": `I love this movie. This film is a milestone in cinematography. Great Immersive camera-work. This film is an experience and i has already seen it 4 times and I only see more quality of the film.`,
  "date": `2020-10-09T13:38:44.769Z`
}, {
  "id": 2,
  "user": {
    "id": 25,
    "name": `Sophie`
  },
  "rating": 9,
  "comment": `This movie is just plain bad. There must be some big payola going round this awards season. Badly written, average acting at best, all the characters are unrelatable and inlikeable. 2 hours of my life wasted.`,
  "date": `2020-11-04T13:38:44.769Z`
}, {
  "id": 3,
  "user": {
    "id": 7,
    "name": `Jack`
  },
  "rating": 9,
  "comment": `Unfortunately we don't have a reliable way to tell the true ratings of a movie.`,
  "date": `2020-10-09T13:38:44.769Z`
}];

const avatarUrl = `/avatar.jpg`;

const user = {
  "id": 1,
  "email": `Oliver.conner@gmail.com`,
  "name": `Oliver.conner`,
  "avatar_url": `img/1.png`
};

const noop = () => {};

const genres = [`Comedy`, `Drama`, `Thriller`];

export {films, genres, reviews, avatarUrl, user, film, filmsRaw, noop};
