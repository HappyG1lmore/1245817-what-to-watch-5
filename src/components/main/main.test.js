import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import {Main} from './main';
import {films, genres, noop} from '../../test-data';

const renderer = new ShallowRenderer();

test(`Main renders correctly`, () => {
  renderer.render(
      <Main
        filteredFilms={films}
        filmsGenres={genres}
        history={{push: noop}}
        mainFilm={films[1]}
      />
  );
  const tree = renderer.getRenderOutput();
  expect(tree).toMatchSnapshot();
});
