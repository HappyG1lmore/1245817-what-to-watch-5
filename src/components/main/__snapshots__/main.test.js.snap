// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main renders correctly 1`] = `
<React.Fragment>
  <section
    className="movie-card"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="Pulp Fiction"
        src="https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <Memo(Connect(Header))
      className="movie-card__head"
    />
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster"
        >
          <img
            alt="The Grand Budapest Hotel poster"
            height="327"
            src="https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            Pulp Fiction
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Crime
            </span>
            <span
              className="movie-card__year"
            >
              1994
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <Memo(Connect(FavoriteButton))
              id={2}
              isFavorite={false}
            />
          </div>
        </div>
      </div>
    </div>
  </section>
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ul
        className="catalog__genres-list"
      >
        <Memo(Connect(GenresList))
          filmGenres={
            Array [
              "Comedy",
              "Drama",
              "Thriller",
            ]
          }
        />
      </ul>
      <div
        className="catalog__movies-list"
      >
        <FilmsList
          filmsList={
            Array [
              Object {
                "background": "https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg",
                "backgroundColor": "#73B39A",
                "description": "A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.",
                "director": "Nicolas Winding Refn",
                "frame": "https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg",
                "genre": "Action",
                "id": 1,
                "isFavorite": false,
                "poster": "https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg",
                "previewVideoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
                "rating": 3.6,
                "ratingText": "Normal",
                "ratings": "109661",
                "runtime": 92,
                "starring": Array [
                  "Tom Hardy",
                  "Kelly Adams",
                  "Luing Andrews",
                ],
                "title": "Bronson",
                "video": "http://media.xiph.org/mango/tears_of_steel_1080p.webm",
                "year": "2008",
              },
              Object {
                "background": "https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg",
                "backgroundColor": "#795433",
                "description": "The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.",
                "director": "Quentin Tarantino",
                "frame": "https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg",
                "genre": "Crime",
                "id": 2,
                "isFavorite": false,
                "poster": "https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg",
                "previewVideoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
                "rating": 1.5,
                "ratingText": "Bad",
                "ratings": "1635992",
                "runtime": 153,
                "starring": Array [
                  "John Travolta",
                  "Uma Thurman",
                  "Samuel L. Jackson",
                ],
                "title": "Pulp Fiction",
                "video": "http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4",
                "year": "1994",
              },
              Object {
                "background": "https://assets.htmlacademy.ru/intensives/javascript-3/film/background/matrix.jpg",
                "backgroundColor": "#B9B27E",
                "description": "A computer hacker learns from mysterious rebels about the true nature of his reality and his role in the war against its controllers.",
                "director": "Wachowski Brothers",
                "frame": "https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/matrix.jpg",
                "genre": "Action",
                "id": 4,
                "isFavorite": false,
                "poster": "https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/matrix.jpg",
                "previewVideoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
                "rating": 4.4,
                "ratingText": "Normal",
                "ratings": "1503092",
                "runtime": 136,
                "starring": Array [
                  "Keanu Reeves",
                  "Laurence Fishburne",
                  "Carrie-Anne Moss",
                ],
                "title": "Matrix",
                "video": "http://media.xiph.org/mango/tears_of_steel_1080p.webm",
                "year": "1999",
              },
            ]
          }
        />
      </div>
      <MoreFilmsButton
        isVisible={false}
        onClick={[Function]}
      />
    </section>
    <Footer />
  </div>
</React.Fragment>
`;
